File: Lab Final/src/main/java/org/example/command/ChangeIssueStatusCommand.java
package org.example.command;

import org.example.model.Issue;
import org.example.model.Status;
import org.example.repository.IssueRepository;

public class ChangeIssueStatusCommand extends AbstractCommand implements Command {
    private final String issueId;
    private final Status newStatus;
    private final Status oldStatus;
    private final IssueRepository repository;

    public ChangeIssueStatusCommand(String issueId, Status newStatus, IssueRepository repository) {
        super(CommandType.UPDATE);
        this.issueId = issueId;
        this.newStatus = newStatus;
        this.repository = repository;

        Issue issue = repository.findById(issueId)
                .orElseThrow(() -> new IllegalArgumentException("Issue not found: " + issueId));
        this.oldStatus = issue.getStatus();
    }

    public void execute() {
        repository.findById(issueId).ifPresent(issue -> {
            issue.setStatus(newStatus);
            repository.save(issue);
        });
    }

    public void undo() {
        repository.findById(issueId).ifPresent(issue -> {
            issue.setStatus(oldStatus);
            repository.save(issue);
        });
    }

    public String getDescription() {
        return "Changed issue status from " + oldStatus + " to " + newStatus + " for issue: " + issueId;
    }

    public String getIssueId() {
        return issueId;
    }

    public Status getNewStatus() {
        return newStatus;
    }

    public Status getOldStatus() {
        return oldStatus;
    }
}

File: LabFinal/TaskManager/TaskManager/Commands/ChangeIssueStatusCommand.cs
using TaskManager.Domain;
using TaskManager.Repository;

namespace TaskManager.Commands;

public class ChangeIssueStatusCommand : ICommand
{
    private readonly IssueRepository _repository;
    private readonly int _issueId;
    private readonly IssueStatus _newStatus;
    private IssueStatus _oldStatus;

    public ChangeIssueStatusCommand(IssueRepository repository, int issueId, IssueStatus newStatus)
    {
        _repository = repository;
        _issueId = issueId;
        _newStatus = newStatus;
        Timestamp = DateTime.Now;
    }

    public void Execute()
    {
        var issue = _repository.GetById(_issueId);
        _oldStatus = issue.Status;
        issue.Status = _newStatus;
        issue.LastModifiedDate = DateTime.Now;
        _repository.Update(issue);
    }

    public void Undo()
    {
        var issue = _repository.GetById(_issueId);
        issue.Status = _oldStatus;
        _repository.Update(issue);
    }

    public string Description => $"Changed issue #{_issueId} status to {_newStatus}";
    public DateTime Timestamp { get; }
    public CommandType Type => CommandType.Update;
}